### DISCUSSION TITLE
Definition of dh_dq in computeCentroidalDynamicsDerivatives() when root is floating

### DISCUSSION THREAD

**Discussion by HiroIshida** [stack-of-tasks/pinocchio#2260] [Q&A]
Definition of dh_dq in computeCentroidalDynamicsDerivatives()

In the documentation, `dh_dq` is explained as
```
The partial derivative of the centroidal momentum with respect to the configuration vector (dim 6 x model.nv).
```
https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#a344d8d1d4380d93f228a880b0c3361fc

Thing that confused me is that in pinocchio the configuration vector is of size `model.nq` > `model.nv` in floating robot because confugration is represented using quaternion which has 1dim redundancy. Could you please explain the mathematical definition of `dh_dq` when root link is floating?

One possibility I guess is that $J$ =dh_dq is a matrix in relation $\dot{h} = J [v^T, \omega^T, q_j^T]^T$ where $v \in \mathbb{R}^3$ and $\omega \in \mathbb{R}^3$ are linear and angular velocity of the root link, respectively and $\dot{q}_j$ is the joint velocities. I s this correct?

**Comment by jorisv** [comment]
Hello @HiroIshida,

You're correct.
In Pinocchio $q \in{R^{nq}}$ (generalized position vector) and $\dot{q}\in{R^{nv}}$ (generalized velocity vector) can have different size.
All jacobians can be multiplied against a generalized velocity vector to compute the associated jacobian value.

By example, $v_i = J_i \dot{q}$ where $v_i$ is the body $i$ spatial velocity and $J_i$ is the body $i$ jacobian.

Then, in your case $\dot{h} = J_h \dot{q}$ where $J_h = \frac{dh}{dq}$ is correct.

Be careful that $q$ and $\dot{q}$ should contains the freeflyer configuration and velocity as the freeflyer joint is considered as a standard joint in Pinocchio.
